### PR TITLE
Support for Monolog 2

### PR DESCRIPTION
**Pull request by derrabus**
Allows installing Monolog 2 as soon as symfony/symfony#27905 is merged.

### CODE REVIEW

**Comment by Seldaek** [comment]
I wonder if this should error if monolog-bridge is in version <5 or not.. Otherwise it might be confusing if Composer lets you update to monolog 2 but then it fails to run.. Maybe worth making a new major release of the bundle for monolog2 + symfony5 support? Thoughts @lyrixx?

Although I guess the monolog-bridge 4.x will prevent installing monolog 2.x and that fixes it? Except for people using symfony/symfony still.

**Comment by lyrixx** [comment]
I'm a bit lost :/ Is this possible to have every version work with every other version?
If not (or too hard), I'm fine with a next major release

**Comment by Seldaek** [comment]
I think it's possible to have symfony5 + monolog1, but not symfony3/4 + monolog2 as the bridge is not going to be made compatible as far as I understand?

**Comment by derrabus** [comment]
> Although I guess the monolog-bridge 4.x will prevent installing monolog 2.x and that fixes it?

Yes, MonologBridge 4.4 and below depends on Monolog 1 and thus will block the installation of Monolog 2.

> Except for people using symfony/symfony still.

That practice is discouraged for over two years now, but yes: Composer won't prevent the installation of Monolog 2 in that case. Anything we can do to detect that case?

**Comment by derrabus** [comment]
> I think it's possible to have symfony5 + monolog1, but not symfony3/4 + monolog2 as the bridge is not going to be made compatible as far as I understand?

If my PR is merged as it is right now, that would be the plan.

**Comment by Seldaek** [comment]
Sounds fine to me to merge as is for now then and avoid a major version of the bundle.

**Comment by derrabus** [comment]
> Except for people using symfony/symfony still.

On second thought: This might still be true for Symfony 3.4 applications. In Symfony 3, it was still best practice to depend on the monorepo.

https://github.com/symfony/symfony-standard/blob/ffb00c045b980f8f721ddf5531c685bfcd79130c/composer.json#L26

I'll try to build some detection logic for that case.

**Comment by derrabus** [comment]
I've added a check for the monorepo case. Composer cannot prevent the installation of `symfony/symfony` 3.4 alongside Monolog 2, but at least we can blow up the container compilation in that case. 🙈

**Comment by lyrixx** [comment]
> I've added a check for the monorepo case. Composer cannot prevent the installation of `symfony/symfony` 3.4 alongside Monolog 2, but at least we can blow up the container compilation in that case.

Why not using the `conflict` directive of composer?

**Comment by derrabus** [comment]
> Why not using the `conflict` directive of composer?

🤔 That could work as well. 

In `symfony/symfony`:
```JSON
"conflict": {
    "monolog/monolog": ">=2"
}
```
If we do this now, Symfony 3.4.30, 4.2.11 and 4.3.3 will have that conflict rule. Then we change composer.json of this PR into
```JSON
"require": {
    "symfony/monolog-bridge": "^3.4.30|~4.2.11|^4.3.3|^5.0"
}
```
and we should be fine. Or am I thinking too complicated? 😅

**Comment by derrabus** [comment]
See symfony/symfony#32469. If that PR gets merged, I will revert my changes to the extension.

**Comment by Seldaek** [comment]
I think I'd merge as is.

**Comment by inverse** [comment]
What's the best way to start working on adding support for new handlers that'll come in the v2 series?

**Comment by Seldaek** [comment]
I think we should first merge this (any objection @lyrixx ? And then could add support for new handlers, but in the Extension class it should check that `Monolog\Logger::API === 2` before allowing using them.

File: DependencyInjection/MonologExtension.php [hunk 1]
```diff
@@ -48,6 +51,10 @@ private function levelToMonologConst($level)
      */
     public function load(array $configs, ContainerBuilder $container)
     {
+        if (class_exists(FullStack::class) && Kernel::MAJOR_VERSION < 5 && Logger::API >= 2) {
```
Comment (lyrixx) [inline DependencyInjection/MonologExtension.php]: Why this check on `FullStack` ?

**Comment by versh23** [comment]
monolog 2 has been released

File: DependencyInjection/MonologExtension.php (hunk 1 above)
Comment (derrabus) [reply DependencyInjection/MonologExtension.php]: I want to detect that case that Monolog 2 is installed together with `symfony/symfony` 2.x/3.x/4.x. I don't see a good way to make Composer prevent that combination, so I'm catching that case here. If you know a better way, I'm open for suggestions.

Comment (Seldaek) [reply DependencyInjection/MonologExtension.php]: I think the point is the Kernel::MAJOR_VERSION is enough? Or why do you need three datapoints to check two versions?

Comment (derrabus) [reply DependencyInjection/MonologExtension.php]: `Kernel::MAJOR_VERSION` is the version of the HttpKernel component, not the version of MonologBridge. But it's the bridge that is actually incompatible before version 5. I have no way to check for the bridge version programmatically. The assumption that the versions of MonologBridge and HttpKernel match is only valid if we have the monorepo.

If I removed the monorepo check, we would also throw if the app uses the standalone packages HttpKernel 4.4, MonologBridge 5.0 and Monolog 2, but that combination should work fine.

Comment (Seldaek) [reply DependencyInjection/MonologExtension.php]: Ok gotcha, then sounds good to me to merge as is.. @lyrixx ?

Comment (lyrixx) [reply DependencyInjection/MonologExtension.php]: Why do we allow symfony/symfony 4.4 and monolog/monolog 2 in composer, but throw an exception manually? 
I mean, We could forbid installation thanks to composer, isn't ?

Comment (Seldaek) [reply DependencyInjection/MonologExtension.php]: That was turned down in https://github.com/symfony/symfony/pull/32469 - IMO this is fine, people using the bundle with the wrong version of symfony+monolog will get an exception, then they can fix.. Not ideal but not so bad either..

Other alternative is to require Monolog 2 and Symfony 5 or at least the monolog bridge in v5 then we fork the bundle completely in two branches for monolog 1 and 2 support.

Comment (lyrixx) [reply DependencyInjection/MonologExtension.php]: :+1:

**Review by lyrixx — APPROVED** [review]

**Pull request merged**